application/server/server.go

```diff
@@ -0,0 +1,174 @@
+// handles requests/responses and their encoding/decoding.
+// A ConiksServer also supports concurrent handling of requests and
+// a mechanism to update the underlying ConiksDirectory automatically
+// at regular time intervals.
+type ConiksServer struct {
+	*application.ServerBase
+	dir        *directory.ConiksDirectory
+	epochTimer *time.Timer
+}
+
+var _ application.Server = (*ConiksServer)(nil)
+
+// NewConiksServer creates a new reference implementation of
+// a CONIKS key server.
+func NewConiksServer(conf *Config) *ConiksServer {
+	// determine this server's request permissions
+	perms := make(map[*application.ServerAddress]map[int]bool)
+
+	for i := 0; i < len(conf.Addresses); i++ {
+		addr := conf.Addresses[i]
+		perms[addr.ServerAddress] = make(map[int]bool)
+		perms[addr.ServerAddress][protocol.KeyLookupType] = true
+		perms[addr.ServerAddress][protocol.KeyLookupInEpochType] = true
+		perms[addr.ServerAddress][protocol.MonitoringType] = true
+		perms[addr.ServerAddress][protocol.RegistrationType] = addr.AllowRegistration
+	}
+
+	// create server instance
+	sb := application.NewServerBase(conf.ServerBaseConfig, "Listen",
+		perms)
+
+	server := &ConiksServer{
+		ServerBase: sb,
+		dir: directory.New(
+			conf.Policies.EpochDeadline,
+			conf.Policies.vrfKey,
+			conf.Policies.signKey,
+			conf.LoadedHistoryLength,
+			true),
+		epochTimer: time.NewTimer(time.Duration(conf.Policies.EpochDeadline) * time.Second),
+	}
+
+	return server
+}
+
+// EpochUpdate runs a CONIKS key server's directory epoch update procedure.
+func (server *ConiksServer) EpochUpdate() {
+	server.epochUpdate()
+	server.WaitStopDone()
+}
+
+// ConfigHotReload implements hot-reloading the configuration by
+// listening for SIGUSR2 signal.
+func (server *ConiksServer) ConfigHotReload() {
+	server.updatePolicies()
+	server.WaitStopDone()
+}
+
+// HandleRequests validates the request message and passes it to the
+// appropriate operation handler according to the request type.
+func (server *ConiksServer) HandleRequests(req *protocol.Request) *protocol.Response {
+	switch req.Type {
+	case protocol.RegistrationType:
+		if msg, ok := req.Request.(*protocol.RegistrationRequest); ok {
+			return server.dir.Register(msg)
+		}
+	case protocol.KeyLookupType:
+		if msg, ok := req.Request.(*protocol.KeyLookupRequest); ok {
+			return server.dir.KeyLookup(msg)
+		}
+	case protocol.KeyLookupInEpochType:
+		if msg, ok := req.Request.(*protocol.KeyLookupInEpochRequest); ok {
+			return server.dir.KeyLookupInEpoch(msg)
+		}
+	case protocol.MonitoringType:
+		if msg, ok := req.Request.(*protocol.MonitoringRequest); ok {
+			return server.dir.Monitor(msg)
+		}
+	}
+
+	return protocol.NewErrorResponse(protocol.ErrMalformedMessage)
+}
+
+// Run implements the main functionality of the key server.
+// It listens for all declared connections with corresponding
+// permissions.
+func (server *ConiksServer) Run(addrs []*Address) {
+	server.WaitStopAdd()
+	go server.EpochUpdate()
```

Comment:
I think we could do
```
server.WaitStopAdd()
go func() {
  server.EpochUpdate()
  server.WaitStopDone()
}
```
It looks more idiomatic Go. 

That being said, I would love to find another way to refactor these APIs.

Reply:
I agree, I'm not 100% happy with these APIs. If you have an alternative suggestion, please let me know! We might find a better way to express this API when implementing the auditor.

Reply:
According to places like https://tour.golang.org/concurrency/1, `go server.EpochUpdate()` is just as idiomatic in Go. Is there a reason why these goroutines should be in anonymous functions?

Reply:
Calling for help @Liamsi  @arlolra ...